presto-main/src/main/java/com/facebook/presto/connector/system/KillQueryProcedure.java

```diff
@@ -55,19 +55,20 @@ public void killQuery(String queryId, String message)
         QueryId query = parseQueryId(queryId);
 
         try {
-            QueryInfo queryInfo = queryManager.getQueryInfo(query);
+            Optional<QueryState> state = queryManager.getQueryState(query);
+            if (!state.isPresent()) {
+                throw new NoSuchElementException();
+            }
 
             // check before killing to provide the proper error message (this is racy)
-            if (queryInfo.getState().isDone()) {
+            if (state.get().isDone()) {
                 throw new PrestoException(NOT_SUPPORTED, "Target query is not running: " + queryId);
             }
 
             queryManager.failQuery(query, createKillQueryException(message));
 
             // verify if the query was killed (if not, we lost the race)
-            queryInfo = queryManager.getQueryInfo(query);
-            if ((queryInfo.getState() != QueryState.FAILED) ||
-                    !ADMINISTRATIVELY_KILLED.toErrorCode().equals(queryInfo.getErrorCode())) {
+            if (!ADMINISTRATIVELY_KILLED.toErrorCode().equals(queryManager.getQueryInfo(query).getErrorCode())) {
```

Comment:
This looks like maybe it's a behavior change (could queryManager.failQuery still be executing?), but I think that's fine.